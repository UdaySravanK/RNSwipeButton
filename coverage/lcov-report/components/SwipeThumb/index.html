
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for components/SwipeThumb</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> components/SwipeThumb</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">42.85% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>36/84</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">34.48% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>20/58</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">23.8% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>5/21</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">42.85% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>36/84</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="40" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 40%"></div><div class="cover-empty" style="width: 60%"></div></div>
	</td>
	<td data-value="40" class="pct low">40%</td>
	<td data-value="80" class="abs low">32/80</td>
	<td data-value="34.48" class="pct low">34.48%</td>
	<td data-value="58" class="abs low">20/58</td>
	<td data-value="23.8" class="pct low">23.8%</td>
	<td data-value="21" class="abs low">5/21</td>
	<td data-value="40" class="pct low">40%</td>
	<td data-value="80" class="abs low">32/80</td>
	</tr>

<tr>
	<td class="file high" data-value="styles.js"><a href="styles.js.html">styles.js</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-01-24T02:17:55.689Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    