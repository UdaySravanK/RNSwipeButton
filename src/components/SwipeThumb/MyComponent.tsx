import React, { useCallback } from "react";
import {
  Animated,
  PanResponder,
  PanResponderGestureState,
  Text,
  View,
} from "react-native";

interface Props {
  onStart?: () => void;
  onMove?: () => void;
}

const MyComponent: React.FC<Props> = (props) => {
  const { onStart, onMove } = props;

  function onPanResponderStart() {
    onStart && onStart();
  }

  const onPanMove = useCallback(
    async (_: any, _g: PanResponderGestureState) => {
      _g.dx > 0 && onMove && onMove();
    },
    [],
  );

  const panResponder = useCallback(
    PanResponder.create({
      onStartShouldSetPanResponder: (e: any, s: any) => true,
      onStartShouldSetPanResponderCapture: (e: any, s: any) => true,
      onMoveShouldSetPanResponder: (e: any, s: any) => true,
      onMoveShouldSetPanResponderCapture: (e: any, s: any) => true,
      onShouldBlockNativeResponder: (e: any, s: any) => true,
      onPanResponderGrant: onPanResponderStart,
      onPanResponderMove: onPanMove,
      onPanResponderRelease: (e: any, s: any) => true,
    }) as any,
    [props],
  );

  return (
    <Animated.View {...panResponder.panHandlers} testID="TestAnimView">
      <View>
        <Text>USK</Text>
      </View>
    </Animated.View>
  );
};

export default MyComponent;
