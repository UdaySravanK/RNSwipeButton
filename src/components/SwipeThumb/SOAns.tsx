import React, { useCallback, useState, useEffect, useRef } from "react";
import {
  Animated,
  PanResponder,
  PanResponderGestureState,
  Text,
  View,
} from "react-native";

interface SwipeThumbProps {
  disabled?: boolean;
  onStart?: () => void;
  onMove?: () => void;
}

const SwipeThumb: React.FC<SwipeThumbProps> = React.memo((props) => {
  const { disabled = false, onStart, onMove } = props;

  var defaultContainerWidth = 50;

  const animatedWidth = useRef(
    new Animated.Value(defaultContainerWidth),
  ).current;
  const [defaultWidth, setDefaultWidth] = useState(defaultContainerWidth);

  useEffect(() => {
    Animated.timing(animatedWidth, {
      toValue: defaultWidth,
      duration: 400,
      useNativeDriver: false,
    }).start();
  }, [defaultWidth]);

  function onPanResponderStart() {
    if (disabled) return;
    onStart && onStart();
  }

  const onPanResponderMove = useCallback(
    async (_: any, gestureState: PanResponderGestureState) => {
      console.log(gestureState.dx);
      if (disabled) return;
      setDefaultWidth(gestureState.dx);
      onMove && onMove();
    },
    [],
  );

  function onPanResponderRelease(
    _: any,
    gestureState: PanResponderGestureState,
  ) {
    console.log(gestureState.dx);
  }

  const panResponder = useCallback(
    PanResponder.create({
      onStartShouldSetPanResponder: (e: any, s: any) => true,
      onStartShouldSetPanResponderCapture: (e: any, s: any) => true,
      onMoveShouldSetPanResponder: (e: any, s: any) => true,
      onMoveShouldSetPanResponderCapture: (e: any, s: any) => true,
      onShouldBlockNativeResponder: (e: any, s: any) => true,
      onPanResponderGrant: onPanResponderStart,
      onPanResponderMove: onPanResponderMove,
      onPanResponderRelease: onPanResponderRelease,
    }) as any,
    [props],
  );

  return (
    <Animated.View
      style={[{ width: animatedWidth }]}
      {...panResponder.panHandlers}
      testID="TestAnimatableView"
    >
      <View>
        <Text>USK</Text>
      </View>
    </Animated.View>
  );
});

export default SwipeThumb;
