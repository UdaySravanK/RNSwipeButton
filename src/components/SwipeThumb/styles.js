import { StyleSheet } from 'react-native';

const borderWidth = 3;
const margin = 1;
const maxContainerHeight = 100;
const Styles = StyleSheet.create({
  container: {
    alignItems: 'flex-end',
    alignSelf: 'flex-start',
    borderRadius: maxContainerHeight / 2,
    borderRightWidth: 0,
    borderWidth,
    margin,
  },
  containerRTL: {
    alignItems: 'flex-start',
    alignSelf: 'flex-end',
    borderRadius: maxContainerHeight / 2,
    borderLeftWidth: 0,
    borderWidth,
    margin,
  },
  icon: {
    alignItems: 'center',
    borderRadius: maxContainerHeight / 2,
    borderWidth: 2,
    justifyContent: 'center',
    marginVertical: -borderWidth,
  },
});

export default Styles;
export { borderWidth, margin };
