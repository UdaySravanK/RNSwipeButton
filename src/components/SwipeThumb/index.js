import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import {
  Animated,
  Image,
  PanResponder,
  TouchableNativeFeedback,
  View,
} from 'react-native';

// Styles
import styles, { borderWidth, margin } from './styles';

// Constants
import { TRANSPARENT_COLOR } from '../../constants';

const SwipeThumb = props => {
  const paddingAndMarginsOffset = borderWidth + 2 * margin;
  const defaultContainerWidth = props.iconSize;
  const maxWidth = props.layoutWidth - paddingAndMarginsOffset;

  const animatedWidth = useRef(new Animated.Value(defaultContainerWidth))
    .current;
  const [defaultWidth, setDefaultWidth] = useState(defaultContainerWidth);

  const [backgroundColor, setBackgroundColor] = useState(TRANSPARENT_COLOR);
  const [borderColor, setBorderColor] = useState(TRANSPARENT_COLOR);

  const panResponder = useRef(
    PanResponder.create({
      onStartShouldSetPanResponder: () => true,
      onStartShouldSetPanResponderCapture: () => true,
      onMoveShouldSetPanResponder: () => true,
      onMoveShouldSetPanResponderCapture: () => true,
      onShouldBlockNativeResponder: () => true,
      onPanResponderStart,
      onPanResponderMove,
      onPanResponderRelease,
    }),
  ).current;

  useEffect(() => {
    Animated.timing(animatedWidth, {
      toValue: defaultWidth,
      duration: 400,
      useNativeDriver: false,
    }).start();
  }, [animatedWidth, defaultWidth]);

  function onSwipeNotMetSuccessThreshold() {
    // Animate to initial position
    setDefaultWidth(defaultContainerWidth);
    props.onSwipeFail && props.onSwipeFail();
  }

  function onSwipeMetSuccessThreshold(newWidth) {
    if (newWidth !== maxWidth) {
      finishRemainingSwipe();
      return;
    }
    props.onSwipeSuccess && props.onSwipeSuccess();
    reset();
  }

  function onPanResponderStart() {
    if (props.disabled) {
      return;
    }
    props.onSwipeStart && props.onSwipeStart();
  }

  async function onPanResponderMove(event, gestureState) {
    if (props.disabled) {
      return;
    }
    const rtlMultiplier = props.enableRightToLeftSwipe ? -1 : 1;
    const newWidth = defaultContainerWidth + rtlMultiplier * gestureState.dx;
    if (newWidth < defaultContainerWidth) {
      // Reached starting position
      reset();
    } else if (newWidth > maxWidth) {
      // Reached end position
      setBackgroundColors();
      setDefaultWidth(maxWidth);
    } else {
      setBackgroundColors();
      await Animated.timing(animatedWidth, {
        toValue: newWidth,
        duration: 0,
        useNativeDriver: false,
      }).start();
      setDefaultWidth(newWidth);
    }
  }

  function onPanResponderRelease(event, gestureState) {
    if (props.disabled) {
      return;
    }
    const rtlMultiplier = props.enableRightToLeftSwipe ? -1 : 1;
    const newWidth = defaultContainerWidth + rtlMultiplier * gestureState.dx;
    const successThresholdWidth =
      maxWidth * (props.swipeSuccessThreshold / 100);
    newWidth < successThresholdWidth
      ? onSwipeNotMetSuccessThreshold()
      : onSwipeMetSuccessThreshold(newWidth);
  }

  function setBackgroundColors() {
    const { railFillBackgroundColor, railFillBorderColor } = props;
    // Set backgroundColor only if not already set
    if (backgroundColor === TRANSPARENT_COLOR) {
      setBackgroundColor(railFillBackgroundColor);
      setBorderColor(railFillBorderColor);
    }
  }

  function finishRemainingSwipe() {
    // Animate to final position
    setDefaultWidth(maxWidth);
    props.onSwipeSuccess && props.onSwipeSuccess();

    //Animate back to initial position
    props.shouldResetAfterSuccess && reset();
  }

  function reset() {
    setDefaultWidth(defaultContainerWidth);

    if (backgroundColor !== TRANSPARENT_COLOR) {
      setBackgroundColor(TRANSPARENT_COLOR);
      setBorderColor(TRANSPARENT_COLOR);
    }
  }

  function renderThumbIcon() {
    const {
      disabled,
      disabledThumbIconBackgroundColor,
      disabledThumbIconBorderColor,
      iconSize,
      thumbIconBackgroundColor,
      thumbIconBorderColor,
      thumbIconComponent: ThumbIconComponent,
      thumbIconImageSource,
      thumbIconStyles,
    } = props;
    const dynamicStyles = {
      ...thumbIconStyles,
      height: iconSize,
      width: iconSize,
      backgroundColor: disabled
        ? disabledThumbIconBackgroundColor
        : thumbIconBackgroundColor,
      borderColor: disabled
        ? disabledThumbIconBorderColor
        : thumbIconBorderColor,
      overflow: 'hidden',
    };

    return (
      <View style={[styles.icon, { ...dynamicStyles }]}>
        {!ThumbIconComponent && thumbIconImageSource && (
          <Image resizeMethod="resize" source={thumbIconImageSource} />
        )}
        {ThumbIconComponent && (
          <View>
            <ThumbIconComponent />
          </View>
        )}
      </View>
    );
  }

  const {
    disabled,
    enableRightToLeftSwipe,
    onSwipeSuccess,
    railStyles,
    screenReaderEnabled,
    title,
  } = props;

  const panStyle = {
    backgroundColor,
    borderColor,
    width: animatedWidth,
    ...railStyles,
    ...(enableRightToLeftSwipe ? styles.containerRTL : styles.container),
  };

  return (
    <>
      {screenReaderEnabled ? (
        <TouchableNativeFeedback
          accessibilityLabel={`${title}. ${
            disabled ? 'Disabled' : 'Double-tap to activate'
            }`}
          disabled={disabled}
          onPress={onSwipeSuccess}
          accessible>
          <View style={[panStyle, { width: defaultContainerWidth }]}>
            {renderThumbIcon()}
          </View>
        </TouchableNativeFeedback>
      ) : (
          <Animated.View style={[panStyle]} {...panResponder.panHandlers}>
            {renderThumbIcon()}
          </Animated.View>
        )}
    </>
  );
};

SwipeThumb.defaultProps = {
  disabled: false,
  layoutWidth: 0,
  resetAfterSuccessAnimDuration: 200,
  screenReaderEnabled: false,
  thumbIconStyles: {},
};

SwipeThumb.propTypes = {
  disabled: PropTypes.bool,
  disabledThumbIconBackgroundColor: PropTypes.string,
  disabledThumbIconBorderColor: PropTypes.string,
  enableRightToLeftSwipe: PropTypes.bool,
  iconSize: PropTypes.number,
  layoutWidth: PropTypes.number,
  onSwipeFail: PropTypes.func,
  onSwipeStart: PropTypes.func,
  onSwipeSuccess: PropTypes.func,
  railFillBackgroundColor: PropTypes.string,
  railFillBorderColor: PropTypes.string,
  railStyles: PropTypes.object,
  resetAfterSuccessAnimDuration: PropTypes.number,
  screenReaderEnabled: PropTypes.bool,
  shouldResetAfterSuccess: PropTypes.bool,
  swipeSuccessThreshold: PropTypes.number,
  thumbIconBackgroundColor: PropTypes.string,
  thumbIconBorderColor: PropTypes.string,
  thumbIconComponent: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.node,
    PropTypes.func,
  ]),
  thumbIconImageSource: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  thumbIconStyles: PropTypes.object,
  title: PropTypes.string,
};

export default SwipeThumb;
