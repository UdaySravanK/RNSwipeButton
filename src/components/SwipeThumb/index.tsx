import React, {
  useCallback,
  useState,
  useEffect,
  useRef,
  ReactElement,
  useMemo,
} from "react";
import {
  I18nManager,
  Animated,
  Image,
  PanResponder,
  View,
  ViewStyle,
  ImageSourcePropType,
  PanResponderGestureState,
} from "react-native";

// Styles
import styles, { borderWidth, margin } from "./styles";

// Constants
import {
  DEFAULT_ANIMATION_DURATION,
  SWIPE_SUCCESS_THRESHOLD,
  TRANSPARENT_COLOR,
} from "../../constants";
const RESET_AFTER_SUCCESS_DEFAULT_DELAY = 1000;

interface SwipeThumbProps {
  disabled?: boolean;
  disableResetOnTap?: boolean;
  disabledThumbIconBackgroundColor?: string;
  disabledThumbIconBorderColor?: string;
  enableReverseSwipe?: boolean;
  finishRemainingSwipeAnimationDuration?: number;
  forceCompleteSwipe?: (forceComplete: () => void) => void;
  forceReset?: (forceReset: () => void) => void;
  layoutWidth?: number;
  onSwipeFail?: () => void;
  onSwipeStart?: () => void;
  onSwipeSuccess?: (isForceComplete: boolean) => void;
  railFillBackgroundColor?: string;
  railFillBorderColor?: string;
  railStyles?: ViewStyle;
  resetAfterSuccessAnimDelay?: number;
  shouldResetAfterSuccess?: boolean;
  swipeSuccessThreshold?: number;
  thumbIconBackgroundColor?: string;
  thumbIconBorderColor?: string;
  thumbIconComponent?: () => ReactElement;
  thumbIconHeight?: number;
  thumbIconImageSource?: ImageSourcePropType | undefined;
  thumbIconStyles?: ViewStyle;
  thumbIconWidth?: number;
}

const SwipeThumb: React.FC<SwipeThumbProps> = React.memo((props) => {
  const {
    disabled = false,
    disableResetOnTap = false,
    disabledThumbIconBackgroundColor,
    disabledThumbIconBorderColor,
    enableReverseSwipe,
    finishRemainingSwipeAnimationDuration = DEFAULT_ANIMATION_DURATION,
    forceCompleteSwipe,
    forceReset,
    layoutWidth = 0,
    onSwipeFail,
    onSwipeStart,
    onSwipeSuccess,
    railFillBackgroundColor,
    railFillBorderColor,
    railStyles,
    resetAfterSuccessAnimDelay,
    shouldResetAfterSuccess,
    swipeSuccessThreshold,
    thumbIconBackgroundColor,
    thumbIconBorderColor,
    thumbIconComponent: ThumbIconComponent,
    thumbIconHeight,
    thumbIconImageSource,
    thumbIconStyles = {},
    thumbIconWidth,
  } = props;

  const paddingAndMarginsOffset = borderWidth + 2 * margin;
  var defaultContainerWidth = 0;
  if (thumbIconWidth == undefined && thumbIconHeight != undefined) {
    defaultContainerWidth = thumbIconHeight;
  } else if (thumbIconWidth != undefined) {
    defaultContainerWidth = thumbIconWidth;
  }
  const maxWidth = layoutWidth - paddingAndMarginsOffset;
  const isRTL = I18nManager.isRTL;

  const animatedWidth = useRef(
    new Animated.Value(defaultContainerWidth),
  ).current;
  const [shouldDisableTouch, disableTouch] = useState(false);

  const [backgroundColor, setBackgroundColor] = useState(TRANSPARENT_COLOR);
  const [borderColor, setBorderColor] = useState(TRANSPARENT_COLOR);

  useEffect(() => {
    forceReset && forceReset(reset);
  }, [forceReset]);

  useEffect(() => {
    forceCompleteSwipe && forceCompleteSwipe(forceComplete);
  }, [forceCompleteSwipe]);

  function updateWidthWithAnimation(newWidth: number) {
    Animated.timing(animatedWidth, {
      toValue: newWidth,
      duration: finishRemainingSwipeAnimationDuration,
      useNativeDriver: false,
    }).start();
  }

  function updateWidthWithoutAnimation(newWidth: number) {
    Animated.timing(animatedWidth, {
      toValue: newWidth,
      duration: 0,
      useNativeDriver: false,
    }).start();
  }

  function onSwipeNotMetSuccessThreshold() {
    // Animate to initial position
    updateWidthWithAnimation(defaultContainerWidth);
    onSwipeFail && onSwipeFail();
  }

  function onSwipeMetSuccessThreshold(newWidth: number) {
    if (newWidth !== maxWidth) {
      // Animate to final position
      finishRemainingSwipe();
      return;
    }
    invokeOnSwipeSuccess(false);
  }

  function onPanResponderStart() {
    if (disabled) {
      return;
    }
    onSwipeStart && onSwipeStart();
  }

  const onPanResponderMove = useCallback(
    (_: any, gestureState: PanResponderGestureState) => {
      if (disabled) return;

      const reverseMultiplier = enableReverseSwipe ? -1 : 1;
      const rtlMultiplier = isRTL ? -1 : 1;
      const newWidth =
        defaultContainerWidth +
        rtlMultiplier * reverseMultiplier * gestureState.dx;

      if (newWidth < defaultContainerWidth) {
        reset();
      } else if (newWidth > maxWidth) {
        setBackgroundColors();
        updateWidthWithoutAnimation(maxWidth);
      } else {
        setBackgroundColors();
        updateWidthWithoutAnimation(newWidth);
      }
    },
    [
      disabled,
      defaultContainerWidth,
      maxWidth,
      isRTL,
      enableReverseSwipe,
      reset,
      setBackgroundColors,
      animatedWidth,
    ],
  );

  function onPanResponderRelease(
    _: any,
    gestureState: PanResponderGestureState,
  ) {
    if (disabled) {
      return;
    }
    const threshold = swipeSuccessThreshold
      ? swipeSuccessThreshold
      : SWIPE_SUCCESS_THRESHOLD;
    const reverseMultiplier = enableReverseSwipe ? -1 : 1;
    const rtlMultiplier = isRTL ? -1 : 1;
    const newWidth =
      defaultContainerWidth +
      rtlMultiplier * reverseMultiplier * gestureState.dx;
    const successThresholdWidth = maxWidth * (threshold / 100);
    newWidth < successThresholdWidth
      ? onSwipeNotMetSuccessThreshold()
      : onSwipeMetSuccessThreshold(newWidth);
  }

  function setBackgroundColors() {
    if (railFillBackgroundColor != undefined) {
      setBackgroundColor(railFillBackgroundColor);
    }
    if (railFillBorderColor != undefined) {
      setBorderColor(railFillBorderColor);
    }
  }

  function finishRemainingSwipe() {
    // Animate to final position
    updateWidthWithAnimation(maxWidth);
    invokeOnSwipeSuccess(false);

    //Animate back to initial position after successfully swiped
    const resetDelay =
      DEFAULT_ANIMATION_DURATION +
      (resetAfterSuccessAnimDelay !== undefined
        ? resetAfterSuccessAnimDelay
        : RESET_AFTER_SUCCESS_DEFAULT_DELAY);
    setTimeout(() => {
      shouldResetAfterSuccess && reset();
    }, resetDelay);
  }

  function invokeOnSwipeSuccess(isForceComplete: boolean) {
    disableTouch(disableResetOnTap);
    onSwipeSuccess && onSwipeSuccess(isForceComplete);
  }

  function reset() {
    disableTouch(false);
    updateWidthWithAnimation(defaultContainerWidth);
  }

  function forceComplete() {
    updateWidthWithAnimation(maxWidth);
    invokeOnSwipeSuccess(true);
  }

  const dynamicStyles: ViewStyle = useMemo(() => {
    const iconWidth = thumbIconWidth ?? thumbIconHeight ?? 0;
    return {
      ...thumbIconStyles,
      height: thumbIconHeight,
      width: iconWidth,
      backgroundColor: disabled
        ? disabledThumbIconBackgroundColor
        : thumbIconBackgroundColor,
      borderColor: disabled
        ? disabledThumbIconBorderColor
        : thumbIconBorderColor,
      overflow: "hidden",
    };
  }, [
    thumbIconWidth,
    thumbIconHeight,
    thumbIconStyles,
    disabled,
    disabledThumbIconBackgroundColor,
    thumbIconBackgroundColor,
    disabledThumbIconBorderColor,
    thumbIconBorderColor,
  ]);

  const renderThumbIcon = useCallback(() => {
    return (
      <View
        style={[styles.icon, { ...dynamicStyles }]}
        testID="DefaultThumbIcon"
      >
        {!ThumbIconComponent && thumbIconImageSource && (
          <Image resizeMethod="resize" source={thumbIconImageSource} />
        )}
        {ThumbIconComponent && (
          <View>
            <ThumbIconComponent />
          </View>
        )}
      </View>
    );
  }, [ThumbIconComponent, thumbIconImageSource, dynamicStyles]);

  const panResponder = useCallback(
    PanResponder.create({
      onStartShouldSetPanResponder: (e: any, s: any) => true,
      onStartShouldSetPanResponderCapture: (e: any, s: any) => true,
      onMoveShouldSetPanResponder: (e: any, s: any) => true,
      onMoveShouldSetPanResponderCapture: (e: any, s: any) => true,
      onShouldBlockNativeResponder: (e: any, s: any) => true,
      onPanResponderGrant: onPanResponderStart,
      onPanResponderMove: onPanResponderMove,
      onPanResponderRelease: onPanResponderRelease,
    }) as any,
    [props], // [disabled, enableReverseSwipe, defaultContainerWidth, maxWidth, setBackgroundColors, animatedWidth],
  );

  const panStyle = {
    backgroundColor,
    borderColor,
    width: animatedWidth,
    ...(enableReverseSwipe ? styles.containerRTL : styles.container),
    ...railStyles,
  };

  return (
    <Animated.View
      style={[panStyle]}
      {...panResponder.panHandlers}
      pointerEvents={shouldDisableTouch ? "none" : "auto"}
      testID="SwipeThumb"
    >
      {renderThumbIcon()}
    </Animated.View>
  );
});

export default SwipeThumb;
