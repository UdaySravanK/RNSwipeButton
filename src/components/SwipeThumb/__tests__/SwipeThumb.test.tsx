import React from "react";
import { render, screen } from "@testing-library/react-native";
import SwipeThumb from "../index";
import {
  RAIL_FILL_BACKGROUND_COLOR,
  RAIL_FILL_BORDER_COLOR,
  SWIPE_SUCCESS_THRESHOLD,
  THUMB_ICON_BACKGROUND_COLOR,
  THUMB_ICON_BORDER_COLOR,
} from "../../../constants";
import { expect } from "@jest/globals";

describe("SwipeThumb Component", () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
    jest.clearAllMocks();
  });

  it("should render correctly with default props", async () => {
    const onSwipeStart = jest.fn();

    render(
      <SwipeThumb
        enableReverseSwipe={false}
        layoutWidth={200}
        railFillBackgroundColor={RAIL_FILL_BACKGROUND_COLOR}
        railFillBorderColor={RAIL_FILL_BORDER_COLOR}
        swipeSuccessThreshold={SWIPE_SUCCESS_THRESHOLD}
        thumbIconBackgroundColor={THUMB_ICON_BACKGROUND_COLOR}
        thumbIconBorderColor={THUMB_ICON_BORDER_COLOR}
        thumbIconHeight={50}
        onSwipeStart={onSwipeStart}
      />,
    );

    expect(screen.toJSON()).toMatchSnapshot();
  });
});
