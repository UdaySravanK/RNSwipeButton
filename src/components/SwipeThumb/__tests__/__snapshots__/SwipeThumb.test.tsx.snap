// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeThumb Component should render correctly with default props 1`] = `
<View
  collapsable={false}
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  pointerEvents="auto"
  style={
    {
      "alignItems": "flex-end",
      "alignSelf": "flex-start",
      "backgroundColor": "#00000000",
      "borderColor": "#00000000",
      "borderRadius": 50,
      "borderRightWidth": 0,
      "borderWidth": 3,
      "margin": 1,
      "width": 50,
    }
  }
  testID="SwipeThumb"
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 50,
          "borderWidth": 2,
          "justifyContent": "center",
          "marginVertical": -3,
        },
        {
          "backgroundColor": "#D27030",
          "borderColor": "#3D797C",
          "height": 50,
          "overflow": "hidden",
          "width": 50,
        },
      ]
    }
    testID="DefaultThumbIcon"
  />
</View>
`;
